libsql-storage/src/lib.rs

```diff
@@ -1,4 +1,413 @@
+// - no lock stealing
+// - write set is kept in mem
+// - txns don't use max_frame_no yet
+// - no savepoints, yet
+// - no multi tenancy, uses `default` namespace
+// - txn can read new frames after it started (since there are no read locks)
+
+#[derive(Clone)]
+pub struct DurableWalManager {
+    lock_manager: Arc<Mutex<LockManager>>,
+}
+
+impl DurableWalManager {
+    pub fn new(lock_manager: Arc<Mutex<LockManager>>) -> Self {
+        Self { lock_manager }
+    }
+}
+
+impl WalManager for DurableWalManager {
+    type Wal = DurableWal;
+
+    fn use_shared_memory(&self) -> bool {
+        trace!("DurableWalManager::use_shared_memory()");
+        false
+    }
+
+    fn open(
+        &self,
+        _vfs: &mut Vfs,
+        _file: &mut libsql_sys::wal::Sqlite3File,
+        _no_shm_mode: std::ffi::c_int,
+        _max_log_size: i64,
+        db_path: &std::ffi::CStr,
+    ) -> Result<Self::Wal> {
+        let db_path = db_path.to_str().unwrap();
+        trace!("DurableWalManager::open(db_path: {})", db_path);
+        // TODO: use the actual namespace uuid from the connection
+        let namespace = "default".to_string();
+        Ok(DurableWal::new(namespace, self.lock_manager.clone()))
+    }
+
+    fn close(
+        &self,
+        wal: &mut Self::Wal,
+        _db: &mut libsql_sys::wal::Sqlite3Db,
+        _sync_flags: std::ffi::c_int,
+        _scratch: Option<&mut [u8]>,
+    ) -> Result<()> {
+        trace!("DurableWalManager::close()");
+        wal.end_read_txn();
+        Ok(())
+    }
+
+    fn destroy_log(&self, _vfs: &mut Vfs, _db_path: &std::ffi::CStr) -> Result<()> {
+        trace!("DurableWalManager::destroy_log()");
+        Ok(())
+    }
+
+    fn log_exists(&self, _vfs: &mut Vfs, _db_path: &std::ffi::CStr) -> Result<bool> {
+        trace!("DurableWalManager::log_exists()");
+        Ok(true)
+    }
+
+    fn destroy(self)
+    where
+        Self: Sized,
+    {
+        trace!("DurableWalManager::destroy()");
+    }
+}
+
+pub struct DurableWal {
+    namespace: String,
+    conn_id: String,
+    client: parking_lot::Mutex<StorageClient<Channel>>,
+    frames_cache: SieveCache<std::num::NonZeroU64, Vec<u8>>,
+    write_cache: BTreeMap<u32, rpc::Frame>,
+    lock_manager: Arc<Mutex<LockManager>>,
+    _runtime: Option<tokio::runtime::Runtime>,
+    rt: tokio::runtime::Handle,
+}
+
+impl DurableWal {
+    fn new(namespace: String, lock_manager: Arc<Mutex<LockManager>>) -> Self {
+        let (_runtime, rt) = match tokio::runtime::Handle::try_current() {
+            Ok(h) => (None, h),
+            Err(_) => {
+                let rt = tokio::runtime::Runtime::new().unwrap();
+                let handle = rt.handle().clone();
+                (Some(rt), handle)
+            }
+        };
+        // connect to external storage server
+        // export LIBSQL_STORAGE_SERVER_ADDR=http://libsql-storage-server.internal:5002
+        let address = std::env::var("LIBSQL_STORAGE_SERVER_ADDR")
```

Comment:
I think we should avoid any env var stuff in these deeper layers and have them be at the main.rs level.